agent/proxycfg/api_gateway.go

```diff
@@ -0,0 +1,492 @@
+					upstreams[listenerKey] = append(upstreams[listenerKey], upstream)
+				}
+
+				upstreamID := NewUpstreamIDFromServiceName(service.ServiceName())
+				seenUpstreamIDs[upstreamID] = struct{}{}
+
+				watchOpts := discoveryChainWatchOpts{
+					id:         upstreamID,
+					name:       service.Name,
+					namespace:  service.NamespaceOrDefault(),
+					partition:  service.PartitionOrDefault(),
+					datacenter: h.stateConfig.source.Datacenter,
+				}
+
+				handler := &handlerUpstreams{handlerState: h.handlerState}
+				if err := handler.watchDiscoveryChain(ctx, snap, watchOpts); err != nil {
+					return fmt.Errorf("failed to watch discovery chain for %s: %w", upstreamID, err)
+				}
+			}
+		}
+
+	case *structs.TCPRouteConfigEntry:
+		snap.APIGateway.TCPRoutes.Set(ref, route)
+
+		for _, service := range route.Services {
+			upstreamID := NewUpstreamIDFromServiceName(service.ServiceName())
+			seenUpstreamIDs[upstreamID] = struct{}{}
+
+			// For each listener, check if this route should bind and, if so, create an upstream.
+			for _, listener := range snap.APIGateway.Listeners {
+				shouldBind := false
+				for _, parent := range route.Parents {
+					if h.referenceIsForListener(parent, listener, snap) {
+						shouldBind = true
+						break
+					}
+				}
+				if !shouldBind {
+					continue
+				}
+
+				upstream := structs.Upstream{
+					DestinationName:      service.Name,
+					DestinationNamespace: service.NamespaceOrDefault(),
+					DestinationPartition: service.PartitionOrDefault(),
+					LocalBindPort:        listener.Port,
+					//IngressHosts:         g.Hosts,
+					// Pass the protocol that was configured on the ingress listener in order
+					// to force that protocol on the Envoy listener.
+					Config: map[string]interface{}{
+						"protocol": "tcp",
+					},
+				}
+
+				listenerKey := APIGatewayListenerKey{Protocol: string(listener.Protocol), Port: listener.Port}
+				upstreams[listenerKey] = append(upstreams[listenerKey], upstream)
+			}
+
+			watchOpts := discoveryChainWatchOpts{
+				id:         upstreamID,
+				name:       service.Name,
+				namespace:  service.NamespaceOrDefault(),
+				partition:  service.PartitionOrDefault(),
+				datacenter: h.stateConfig.source.Datacenter,
+			}
+
+			handler := &handlerUpstreams{handlerState: h.handlerState}
+			if err := handler.watchDiscoveryChain(ctx, snap, watchOpts); err != nil {
+				return fmt.Errorf("failed to watch discovery chain for %s: %w", upstreamID, err)
+			}
+		}
+	default:
+		return fmt.Errorf("invalid type for config entry: %T", resp.Entry)
+	}
+
+	snap.APIGateway.Upstreams = upstreams
+	snap.APIGateway.UpstreamsSet = seenUpstreamIDs
+	//snap.APIGateway.Hosts = TODO
+	snap.APIGateway.HostsSet = true
+
+	// Stop watching any upstreams and discovery chains that have become irrelevant
+	for upstreamID, cancelDiscoChain := range snap.APIGateway.WatchedDiscoveryChains {
```

Comment:
does this code not happen in the upstream handler?

Reply:
Not today. You can see that we have to do the same thing in `handlerIngressGateway` [here](https://github.com/hashicorp/consul/blob/d3aa2bd9c563dfbdf98610ff577bb753d4dba506/agent/proxycfg/ingress_gateway.go#L160-L177). I just changed the variable names a bit to make it easier to understand what's happening